src/lib/datatypes/include/sol-memdesc.h

```diff
@@ -0,0 +1,1089 @@
+     */
+    SOL_MEMDESC_TYPE_STRING,
+    /**
+     * @brief null-terminated C-string (@c const char*).
+     *
+     * Initial content is specified in
+     * struct sol_memdesc::defcontent::s. It may be null.
+     *
+     * By default, strings are NOT duplicated neither freed.
+     * @see SOL_MEMDESC_TYPE_STRING
+     */
+    SOL_MEMDESC_TYPE_CONST_STRING,
+    /**
+     * @brief generic pointer (void *).
+     *
+     * If struct sol_memdesc::pointed_item is non-NULL, it will be
+     * managed as such (malloc/free). Note that the initial value is
+     * still defined as a pointer to the actual contents in struct
+     * sol_memdesc::defcontent::p. If that is non-NULL, then the
+     * pointer is allocated and that one will use defaults specified
+     * in struct sol_memdesc::pointed_item::defcontent, then values
+     * from struct sol_memdesc::defcontent::p is applied on top.
+     *
+     * By default the value is based on
+     * struct sol_memdesc::defcontent::p.
+     *
+     * @see SOL_MEMDESC_TYPE_STRUCTURE
+     */
+    SOL_MEMDESC_TYPE_PTR,
+    /**
+     * @brief structure with internal members.
+     *
+     * This is a recursive type with children described in struct
+     * sol_memdesc::structure_members, an array that is
+     * null-terminated (all element members are zeroed).
+     *
+     * During initialization, each member will be considered according
+     * to its initial value. Then, if
+     * struct sol_memdesc::defcontent::p is non-NULL, it will be
+     * applied on top.
+     */
+    SOL_MEMDESC_TYPE_STRUCTURE,
+    /**
+     * @brief an array with internal members.
+     *
+     * This is a pointer to an array of items that are defined in
+     * struct sol_memdesc::array_item. It will not be touched, you
+     * should manage it yourself with struct sol_memdesc::ops.
+     *
+     * To map a struct sol_vector, use
+     * @c .size=sizeof(struct sol_vector),
+     * And provide a @c .array_item with the description on what is to
+     * be in the element, like a structure or a pointer to one, this
+     * way sol_memdesc_init_defaults() will set
+     * struct sol_vector::elem_size to size of array item.
+     * Then you must provide the following struct sol_memdesc::ops:
+     *
+     *  @li @c init_defaults: set @c elem_size from
+     *      @c sol_memdesc_get_size(desc->array_item).
+     *  @li @c array.get_length: return @c len.
+     *  @li @c array.get_element: proxy return of sol_vector_get().
+     *  @li @c array.resize: if shrinking, remember to call
+     *      @c sol_memdesc_free_content(desc->array_item, it) for
+     *      every item that will be removed, then call
+     *      sol_vector_del_range(). If growing, call
+     *      sol_vector_append_n() and initialze items with
+     *      @c sol_memdesc_init_defaults(desc->array_item, it).
+     *
+     * @see SOL_MEMDESC_OPS_VECTOR and SOL_MEMDESC_OPS_PTR_VECTOR.
+     */
+    SOL_MEMDESC_TYPE_ARRAY
+};
```

Comment:
no as usually they do not change size and you should stick with a fixed version. long is a special case since it can be used to store pointers and other platform dependent features.